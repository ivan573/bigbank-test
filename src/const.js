export const Route = {
  HOME: '/',
  GAME: '/game',
  SHOP: '/shop',
  REPUTATION: '/reputation'
}

export const ModalWindowType = {
  QUEST: 'quest',
  QUEST_RESULT: 'questResult',
  PURCHASE_ITEM: 'purchaseItem',
  PURCHASE_RESULT: 'purchaseResult',
  GAME_OVER: 'gameOver'
}
