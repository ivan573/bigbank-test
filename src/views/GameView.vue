<template>
  <div class="game">
    <div class="game__quests">
      <h2 class="game__quests-title">Available quests</h2>
      <GameQuests
        class="game__quests-list"
        :quests="quests"
        @questClick="handleQuestClick"
      />
    </div>
    <ModalWindow
      v-if="modalWindowType"
      :type="modalWindowType"
      :data="modalWindowData"
      @close="handleModalWindowClose"
      @confirm="handleModalWindowConfirmation"
    />
  </div>
</template>

<script>
import { mapState, mapMutations, mapActions } from 'vuex'
import { Route, ModalWindowType } from '@/const'
import GameQuests from '@/components/GameQuests.vue'
import ModalWindow from '@/components/ModalWindow.vue'

export default {
  name: 'GameView',
  components: {
    GameQuests,
    ModalWindow
  },
  data() {
    return {
      modalWindowType: null
    }
  },
  computed: {
    ...mapState('gameData', {
      gameId: state => state.gameId,
      lives: state => state.lives,
      isDataLoading: state => state.isLoading
    }),
    ...mapState('messages', {
      quests: state => state.messages,
      activeQuest: state => state.activeMessage,
      message: state => state.message,
      areQuestsLoading: state => state.isLoading
    }),
    modalWindowData() {
      switch (this.modalWindowType) {
        case ModalWindowType.QUEST:
          return this.activeQuest
        case ModalWindowType.QUEST_RESULT:
        case ModalWindowType.GAME_OVER:
          return { message: this.message }
        default:
          return null
      }
    }
  },
  created() {
    if ((!this.gameId && !this.isDataLoading) || (!this.quests.length && !this.areQuestsLoading)) {
      this.$router.push({ path: Route.HOME })
    }
  },
  methods: {
    ...mapMutations('messages', ['discardMessageResult', 'updateActiveMessage', 'discardActiveMessage']),
    ...mapActions('messages', ['solveMessage']),
    ...mapActions(['endGame']),
    handleQuestClick(questId) {
      this.updateActiveMessage(questId)
      this.modalWindowType = ModalWindowType.QUEST
    },
    async handleModalWindowConfirmation() {
      await this.solveMessage(this.activeQuest.adId)
      this.discardActiveMessage()
      if (this.lives > 0) {
        this.modalWindowType = ModalWindowType.QUEST_RESULT
        return
      }
      this.modalWindowType = ModalWindowType.GAME_OVER
    },
    handleModalWindowClose(type) {
      this.modalWindowType = null
      switch (type) {
        case ModalWindowType.QUEST:
          this.discardActiveMessage()
          break
        case ModalWindowType.QUEST_RESULT:
          this.discardMessageResult()
          break
        case ModalWindowType.GAME_OVER:
          this.endGame()
          this.$router.push({ path: Route.HOME })
          break
      }
    }
  }
}
</script>

<style lang="scss" scoped>
  .game {
    &__data {
      display: flex;
      justify-content: center;
      border-top: $border;
      border-bottom: $border;
    }

    &__data-item {
      padding: 4px 8px;
      border-left: $border;

      &:last-child {
        border-right: $border;
      }
    }

    &__quests {
      display: flex;
      flex-direction: column;
      align-items: center;
    }

    &__quests-list {
      margin: 0 64px 32px;
      max-width: 900px;
    }
  }
</style>
