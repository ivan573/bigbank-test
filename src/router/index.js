import Vue from 'vue'
import VueRouter from 'vue-router'
import HomeView from '@/views/HomeView.vue'
import GameView from '@/views/GameView.vue'
import ShopView from '@/views/ShopView.vue'
import ReputationView from '@/views/ReputationView.vue'

if (process.env.NODE_ENV !== 'test') {
  Vue.use(VueRouter)
}

export const Route = {
  HOME: '/',
  GAME: '/game',
  SHOP: '/shop',
  REPUTATION: '/reputation'
}

export const routes = [
  {
    path: Route.HOME,
    name: 'home',
    component: HomeView
  },
  {
    path: Route.GAME,
    name: 'game',
    component: GameView
  },
  {
    path: Route.SHOP,
    name: 'shop',
    component: ShopView
  },
  {
    path: Route.REPUTATION,
    name: 'repotation',
    component: ReputationView
  }
]

const router = new VueRouter({
  mode: 'history',
  base: process.env.BASE_URL,
  routes
})

export default router
