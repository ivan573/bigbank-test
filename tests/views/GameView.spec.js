import { mount, shallowMount, createLocalVue } from '@vue/test-utils'
import Vuex from 'vuex'
import VueRouter from 'vue-router'
import { ModalWindowType } from '@/const'
import { Route } from '@/router/index'
import GameView from '@/views/GameView.vue'

describe('GameView.vue', () => {
  const gameId = 'gameId'
  const lives = 3
  const isDataLoading = false
  const quests = [
    {
      adId: 'Ln859bV7',
      expiresIn: 4,
      message: 'Help Menes Rennold to transport a magic turnips to grassland in Everdrone',
      probability: 'Piece of cake',
      reward: 15

    },
    {
      adId: 'lBxlpzxe',
      expiresIn: 5,
      message: 'Help Viriato Ásmundsdóttir to sell an unordinary water on the local market',
      probability: 'Walk in the park',
      reward: 10

    },
    {
      adId: 'bauDDuqQ',
      expiresIn: 3,
      message: 'Help Chingis Chandler to fix their horse',
      probability: 'Quite likely',
      reward: 8
    }
  ]
  const activeQuest = quests[0]
  const message = 'You successfully solved the mission!'
  const areQuestsLoading = false

  const computed = {
    gameId() { return gameId },
    lives() { return lives },
    isDataLoading() { return isDataLoading },
    quests() { return quests },
    activeQuest() { return activeQuest },
    message() { return message },
    areQuestsLoading() { return areQuestsLoading }
  }

  it('matches snapshot', () => {
    const wrapper = mount(GameView, { computed })
    expect(wrapper.element).toMatchSnapshot()
  })

  it('whenever a quest emits questClick event, component updates modal window type to quest and passes quest id to store', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const updateActiveMessage = jest.fn()
    const messages = {
      namespaced: true,
      mutations: { updateActiveMessage }
    }
    const store = new Vuex.Store({ modules: { messages } })

    const wrapper = mount(GameView, {
      localVue,
      store,
      computed
    })
    const questsElement = wrapper.find('.game__quests-list')

    questsElement.vm.$emit('questClick', quests[0].adId)

    expect(wrapper.vm.modalWindowType).toBe(ModalWindowType.QUEST)
    expect(updateActiveMessage).toHaveBeenCalled()
    expect(updateActiveMessage).toHaveBeenCalledWith({}, quests[0].adId)
  })

  it('whenever modal window with quest type emits comfirm event, component calls solving action, discards message and updates window type', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const discardActiveMessage = jest.fn()
    const solveMessage = jest.fn()
    const messages = {
      namespaced: true,
      mutations: { discardActiveMessage },
      actions: { solveMessage }
    }
    const store = new Vuex.Store({ modules: { messages } })

    const modalWindowType = ModalWindowType.QUEST
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(GameView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.game__modal-window')

    await modalWindowElement.vm.$emit('confirm', modalWindowType)

    await wrapper.vm.$nextTick()

    expect(solveMessage).toHaveBeenCalled()
    expect(discardActiveMessage).toHaveBeenCalled()
    expect(wrapper.vm.modalWindowType).toBe(ModalWindowType.QUEST_RESULT)
  })

  it('whenever modal window with quest type emits comfirm event and the quest fails taking last live, component updates modal window type to game over', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const discardActiveMessage = jest.fn()
    const solveMessage = jest.fn()
    const messages = {
      namespaced: true,
      mutations: { discardActiveMessage },
      actions: { solveMessage }
    }
    const store = new Vuex.Store({ modules: { messages } })

    const modalWindowType = ModalWindowType.QUEST
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(GameView, {
      localVue,
      store,
      data,
      computed: {
        ...computed,
        lives() { return 0 }
      }
    })

    const modalWindowElement = wrapper.find('.game__modal-window')

    await modalWindowElement.vm.$emit('confirm', modalWindowType)

    await wrapper.vm.$nextTick()

    expect(solveMessage).toHaveBeenCalled()
    expect(discardActiveMessage).toHaveBeenCalled()
    expect(wrapper.vm.modalWindowType).toBe(ModalWindowType.GAME_OVER)
  })

  it('whenever modal window with quest type emits close event, component updates modal window type and call discard mutation', () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const discardActiveMessage = jest.fn()
    const messages = {
      namespaced: true,
      mutations: { discardActiveMessage }
    }
    const store = new Vuex.Store({ modules: { messages } })

    const modalWindowType = ModalWindowType.QUEST
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(GameView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.game__modal-window')

    modalWindowElement.vm.$emit('close', modalWindowType)

    expect(wrapper.vm.modalWindowType).toBe(null)
    expect(discardActiveMessage).toHaveBeenCalled()
  })

  it('whenever modal window with quest result type emits close event, component updates modal window type and call discard mutation', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)

    const discardMessageResult = jest.fn()
    const messages = {
      namespaced: true,
      mutations: { discardMessageResult }
    }
    const store = new Vuex.Store({ modules: { messages } })

    const modalWindowType = ModalWindowType.QUEST_RESULT
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(GameView, {
      localVue,
      store,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.game__modal-window')

    modalWindowElement.vm.$emit('close', modalWindowType)

    expect(wrapper.vm.modalWindowType).toBe(null)
    expect(discardMessageResult).toHaveBeenCalled()
  })

  it('whenever modal window with game over type emits close event, component calls end game action and redirects to root', async () => {
    const localVue = createLocalVue()
    localVue.use(Vuex)
    localVue.use(VueRouter)

    const endGame = jest.fn()
    const actions = {
      endGame
    }
    const store = new Vuex.Store({ actions })

    const router = new VueRouter()
    router.push(Route.GAME)

    const modalWindowType = ModalWindowType.GAME_OVER
    const data = () => {
      return {
        modalWindowType
      }
    }

    const wrapper = mount(GameView, {
      localVue,
      store,
      router,
      data,
      computed
    })

    const modalWindowElement = wrapper.find('.game__modal-window')

    modalWindowElement.vm.$emit('close', modalWindowType)

    expect(wrapper.vm.modalWindowType).toBe(null)
    expect(endGame).toHaveBeenCalled()
    expect(wrapper.vm.$route.path).toBe(Route.HOME)
  })

  it('if the modal window has the quest type component provides active quest data to render', () => {
    const modalWindowType = ModalWindowType.QUEST
    const data = () => {
      return {
        modalWindowType
      }
    }
    const wrapper = mount(GameView, { data, computed })

    expect(wrapper.vm.modalWindowData).toEqual(wrapper.vm.activeQuest)
  })

  it('if the modal window has the quest result type component provides quest result message to render', () => {
    const modalWindowType = ModalWindowType.QUEST_RESULT
    const data = () => {
      return {
        modalWindowType
      }
    }
    const wrapper = mount(GameView, { data, computed })

    expect(wrapper.vm.modalWindowData).toEqual({ message: wrapper.vm.message })
  })

  it('if the modal window has the game over type component provides quest result message to render', () => {
    const modalWindowType = ModalWindowType.GAME_OVER
    const data = () => {
      return {
        modalWindowType
      }
    }
    const wrapper = mount(GameView, { data, computed })

    expect(wrapper.vm.modalWindowData).toEqual({ message: wrapper.vm.message })
  })

  it('redirects to root path if the game is not initialized', () => {
    const localVue = createLocalVue()
    localVue.use(VueRouter)

    const router = new VueRouter()
    router.push(Route.GAME)

    const wrapper = shallowMount(GameView, {
      localVue,
      router,
      computed: {
        ...computed,
        gameId() { return null }
      }
    })

    expect(wrapper.vm.$route.path).toBe(Route.HOME)
  })
})
