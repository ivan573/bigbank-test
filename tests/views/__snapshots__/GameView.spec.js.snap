// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameView.vue matches snapshot 1`] = `
<div
  class="game"
>
  <div
    class="game__quests"
  >
    <h2
      class="game__quests-title"
    >
      Available quests
    </h2>
     
    <ul
      class="game-quests game__quests-list"
    >
      <li
        class="game-quests__quest"
      >
        <p
          class="game-quests__quest-text"
        >
          Help Menes Rennold to transport a magic turnips to grassland in Everdrone
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Reward: 15
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Difficulty: Piece of cake
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Turns before expiration: 4
        </p>
      </li>
      <li
        class="game-quests__quest"
      >
        <p
          class="game-quests__quest-text"
        >
          Help Viriato Ásmundsdóttir to sell an unordinary water on the local market
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Reward: 10
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Difficulty: Walk in the park
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Turns before expiration: 5
        </p>
      </li>
      <li
        class="game-quests__quest"
      >
        <p
          class="game-quests__quest-text"
        >
          Help Chingis Chandler to fix their horse
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Reward: 8
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Difficulty: Quite likely
        </p>
         
        <p
          class="game-quests__quest-text"
        >
          Turns before expiration: 3
        </p>
      </li>
    </ul>
  </div>
   
  <!---->
</div>
`;
